akka-stream/src/main/scala/akka/stream/javadsl/KillSwitch.scala

```diff
@@ -0,0 +1,144 @@
+   */
+  def single[T]: Flow[T, T, UniqueKillSwitch] = singleKillSwitchFlow.asInstanceOf[Flow[T, T, UniqueKillSwitch]]
+
+  private val singleKillSwitchFlow: Flow[Any, Any, UniqueKillSwitch] =
+    Flow.fromGraph(scaladsl.KillSwitch.single[Any].mapMaterializedValue(new UniqueKillSwitch(_)))
+
+  /**
+   * Creates a new [[BidiFlow]] that materializes to an external switch that allows external completion of that unique
+   * materialization. Different materializations result in different, independent switches.
+   *
+   * For a Flow version see [[KillSwitch#single]]
+   */
+  def singleBidi[T1, T2]: BidiFlow[T1, T1, T2, T2, UniqueKillSwitch] =
+    singleKillSwitchBidiFlow.asInstanceOf[BidiFlow[T1, T1, T2, T2, UniqueKillSwitch]]
+
+  private val singleKillSwitchBidiFlow: BidiFlow[Any, Any, Any, Any, UniqueKillSwitch] =
+    BidiFlow.fromGraph(scaladsl.KillSwitch.singleBidi[Any, Any].mapMaterializedValue(new UniqueKillSwitch(_)))
+
+}
+
+/**
+ * A [[UniqueKillSwitch]] is always a result of a materialization (unlike [[SharedKillSwitch]] which is constructed
+ * before any materialization) and it always controls that graph and stage which yielded the materialized value.
+ *
+ * After calling [[UniqueKillSwitch#completeAndCancel()]] the running instance of the [[Flow]] that materialized to the
+ * [[UniqueKillSwitch]] will complete its downstream and cancel its upstream (unless if finished or failed already in which
+ * case the command is ignored). Subsequent invocations of completion commands will be ignored.
+ *
+ * After calling [[UniqueKillSwitch#failAndCancel()]] the running instance of the [[Flow]] that materialized to the
+ * [[UniqueKillSwitch]] will fail its downstream with the provided exception and cancel its upstream
+ * (unless if finished or failed already in which case the command is ignored). Subsequent invocations of
+ * completion commands will be ignored.
+ *
+ * It is also possible to individually cancel, complete or fail upstream and downstream parts by calling the corresponding
+ * methods.
+ */
+final class UniqueKillSwitch private[stream] (private val wrapped: scaladsl.UniqueKillSwitch) {
+  /**
+   * After calling [[UniqueKillSwitch#shutdown()]] the running instance of the [[Flow]] that materialized to the
+   * [[UniqueKillSwitch]] will complete its downstream and cancel its upstream (unless if finished or failed already in which
+   * case the command is ignored). Subsequent invocations of completion commands will be ignored.
+   */
+  def shutdown(): Unit = wrapped.shutdown()
+
+  /**
+   * After calling [[UniqueKillSwitch#abort()]] the running instance of the [[Flow]] that materialized to the
+   * [[UniqueKillSwitch]] will fail its downstream with the provided exception and cancel its upstream
+   * (unless if finished or failed already in which case the command is ignored). Subsequent invocations of
+   * completion commands will be ignored.
+   */
+  def abort(reason: Throwable): Unit = wrapped.abort(reason)
```

Comment:
the stage infrastructure is idempotent in the close events, so thankfully nothing.